Updates the behavior of the top toolbar fixed setting

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #40450
Updates the behavior of the top toolbar fixed setting.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- wraps the edit post header in a slot fill provider
- implements a temporary behavior that shows document tools when no block is selected
- updates the block tools and parent selector to appear properly in the new place
- implements a utility component that shows the document tools when they receive focus via toolbar

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Enable the Top Toolbar setting from the top right dot menu
2. Select a block
3. Notice the block tools (toolbar) replaces the document tools (toolbar)

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

N/A

## Screenshots or screencast <!-- if applicable -->


https://user-images.githubusercontent.com/107534/225633841-3fdd191f-795e-424c-aeca-762443b74c6d.mp4

draganescu committed:
Updates the behavior of the top toolbar fixed setting:

- wraps the edit post header in a slot fill provider
- implements a temporary behavior that shows document tools when no block is selected
- updates the block tools and parent selector to appear properly in the new place

joedolson commented:
I think my biggest concern about this change is that you lose access to a bunch of tools if a block is selected. As a workflow, a common pattern for me on long documents is "Edit Block > Document Overview > Select Block > Edit Block"; and that workflow isn't available to me with this model without first removing focus from a block. That's easy to do with a mouse, but less trivial or discoverable from the keyboard. 

One thing that might help with this would be if there is a way to deselect the block while in the top toolbar edit tools, so that if you've navigated all the way back to the editing tools, you can get back to the top toolbar immediately. Alternately, the edit post header controls could be placed in a popover at the beginning of the toolbar edit tools, so that those controls are collapsed rather than removed.

draganescu commented:
@joedolson in the [final shape of this](https://user-images.githubusercontent.com/846565/210546838-3ea8ba01-6391-4024-8421-cc8e9fd3cd03.png) there is a button next to the block tools on the left (an up arrow) than can be used to go back to document tools, and there is a button on the right (a button with the block icon, a paragraph in the design) to go back to block tools, all while having a block selected. This PR is just not there yet.

Do you think the toggle solution proposed by the design is good enough?

joedolson commented:
I feel like the toggle to switch between menus should stay in the same position. Effectively, the arrow control to switch to the document tools and the paragraph tool to switch to the block tools are just a single switch that changes between toolbars, and having one of those tools at the start of the toolset and the other at the end feels strange to me. I think it would be more predictable if the switch control was always in the same position. But yes, that would help a lot!

draganescu committed:
disables auto toggling to top toolbar in DFM and removes superfluous slot fill provider

aaronrobertshaw committed:
Show block toolbar for multi-block selections

aaronrobertshaw commented:
Hope you don't mind @draganescu, I pushed a commit to this branch to help me test a few other use cases for the Top Toolbar. As of https://github.com/WordPress/gutenberg/pull/47277/commits/06b21538df6e13399cdba1280fd17cf5a0a6b68c the block toolbar will be shown in the top toolbar when multiple blocks are selected. 

| Before | After |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/216514610-01b9039d-3a74-4262-955a-0f247559fa08.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/216514670-a62bf78c-80e4-4a4d-a47d-6f2714571edc.mp4" /> |

alexstine commented:
I am not in favor of this change because of context issues. Let me take a moment and paint some thoughts.
1. When a user opens a post/page and starts to navigate the editor, they will pass through the document tools toolbar.
2. Once they enter the content and focus a block, pressing shift+tab will land in the block toolbar. No change yet.
3. Pressing shift+tab again, what will happen? Will I land on an options button or will I land on a show document tools button?
4. We'll have to take extra steps to manage focus with a button that shows a toolbar.
5. Disappearing and appearing DOM elements are always tricky to communicate to non-sighted users. The show document tools button would need to disappear or have some type of state value.
6. If the show document tools button persists with some type of state, that adds yet another tab stop.
7. For screen reader users who are navigating in virtual mode, there is now no longer a way to access the document tools toolbar since it will be hidden in the DOM/conditionally removed.
8. This further continues the problem in Gutenberg with everything needed to be overly communicated for screen reader users.
9. There are important buttons in the toolbar such as the block inserter and document overview/list view. Hiding these should be a non-starter.

Just my thoughts without testing. Right, wrong, not even close?

Thanks.

talldan commented:
> Just my thoughts without testing. Right, wrong, not even close?

I think you're right on the money that it's going to be difficult to get this feature right. It isn't the default option though, so one good thing is you shouldn't ever need to enable it. (though I realise it should also work in a good way anyway even when enabled.)

One of the main thing that stands out to me is that the tab order is going to be hard to get right. Currently users expect to be able to shift tab from the block to the block toolbar directly without passing through Options, Settings, Publish and Preview buttons. But this change will put those buttons in the way, and that won't be very convenient.

It's worth noting that plugins can also add buttons to this area, which means there could be quite a lot of tab stops.

draganescu committed:
try maing the inserter always visible

alexstine commented:
Here is an idea. How about you visually hide the document tools toolbar and then add events to show it when focus passes through it? That way the experience does not change for keyboard users but you still get the appear/disappear functionality. I simply cannot support adding another tab stop to this area. I was also unaware that there was a slot fill in that area. That really could turn messy one day.

draganescu commented:
> visually hide the document tools toolbar and then add events to show it when focus passes through it

@alexstine I don't grasp how that would work? Does it mean for keyboard navigation both toolbars will be available at the same time?

alexstine commented:
@draganescu Correct. Then when the toolbar receives focus, it will be shown visually. Completely removing or hiding the first toolbar from the DOM I do not think is a good idea.

draganescu commented:
Doing some more research:

- to enable the document tools or block tools to be visually hidden and be shown when they receive focus, but at the same time achieve the design where the inserter is present at all time, a new component that only handles the inserter is required
- ~~for some reason when the block toolbar is rendered in the header chrome it loses the roving behavior and arrows can't be used anymore to move selection of the toolbar~~
  - this is solved by wrapping the block toolbar in a navigable toolbar
- unlike the current fixed toolbar setting, this implementation loses the ability to quickly move from the block to its toolbar using shift + tab. Here shift + tab moves to the document more menu and one has to shift + tab all the way to the block toolbar.

All these need to be tackled to make this a viable direction.

draganescu committed:
visually hide the toolbars when they should be hidden

draganescu committed:
add the roving behavior back to the block tools

draganescu committed:
potential solution for keyboard navigation

draganescu committed:
keep the inserter button around when switching editor toolbars

draganescu committed:
modify the parent selector UI for when the block toolbar is fixed

draganescu commented:
I have updated this with a few things - visible in the video of the PR's description:

- added a utility component that would handle showing document tools when they receive focus via keyboard navigation
- added some tweaks to the parent selector when the toolbar is in the editor header to conform to the UI's design in the parent issue

Immediately next, I need to:

- ~~figure out how to handle the inserter button which now shows twice if both toolbars are visible.~~
- ~~see what's up with the bug that appears in navigation mode which appears to scramble what happens in the editor header with toolbars (visible at the end of the description video)~~

draganescu committed:
hide the block tools inserter when the document tools are focused

draganescu committed:
adds toggling between toolbars, fixes navigation mode phantom buttons

draganescu committed:
adds block icon to the block tools toggle

draganescu commented:
New updates:

- added toggles for mouse users to use for moving between the two sets of tools.

Also,

- I've fixed the ghose buttons appearing in Navigation mode
- There is only one inserter showing up even when the focused document tools show up


Here is the current state:


https://user-images.githubusercontent.com/107534/228312083-5f4d299e-3481-4103-9cc3-0434fb9c18ba.mp4

@jasmussen @mtias @jameskoster take this for a sping and see how the mechanics work. Should there by any animation? If the mechanics prove good, I'd love to merge and iterate in tweaking, as the amount of changes is significant already.

@alexstine I've done some tweaks:

- in this version the document tools remain in the DOM
- selecting the header region and tabbing once lands in the document tools which are accessible via arrow keys like any toolbar
- when top toolbar setting is on tabbing out of the document tools lands in block tools, if a block is selected
- the block toolbar is _not_ the closest toolbar anymore (cc @jasmussen @tellthemachines this may be an issue) because its position in the DOM has changed. I tried to mitigate this by plopping the block toolbar at the end of the header markup, and position it visually with css, but then the hidden visually document tools break the layout when they're in focus.

I think it's coming together good, but we still need to tweak some of the interactions.

Known UI bug: the icon to toggle block tools appears for a split second when selecting a block, it's a bit jarring.

jameskoster commented:
Thanks for the ping, it's definitely coming together!

I don't think it's essential, but the animation [here](https://github.com/WordPress/gutenberg/issues/40450#issuecomment-1344194576) might be nice to try. It could help connect the behaviour, but it might also be distracting. Hard to say for sure, and perhaps fine to try in a follow-up.

---

Here's the svg for the 'up' arrow:

```
<svg fill="none" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
  <path d="m13.53 8.47-1.06 1.06-2.72-2.72V12h-1.5V6.81L5.53 9.53 4.47 8.47 9 3.94l4.53 4.53Zm-1.802 7.968c1.307.697 3.235.812 5.772.812v1.5c-2.463 0-4.785-.085-6.478-.988a4.721 4.721 0 0 1-2.07-2.13C8.48 14.67 8.25 13.471 8.25 12h1.5c0 1.328.208 2.28.548 2.969.332.675.81 1.138 1.43 1.47Z" />
</svg>
```

This icon button (and the block button when the document tools are visible) don't have the same hover / focus states as the other buttons. That would be nice to fix.

---

A nitpick: the separators don't quite match the designs in the issue, they should be the same height as the button footprint but are bit taller at the moment. There should also be separators between the toggle buttons and the toolbars:

<img width="285" alt="Screenshot 2023-03-28 at 18 20 00" src="https://user-images.githubusercontent.com/846565/228318563-0e334178-c02a-4b4f-8419-7c47dd931fc9.png">

---

In the following tab state, the block button is duplicated, and the first one doesn't do anything:

<img width="683" alt="Screenshot 2023-03-28 at 18 27 59" src="https://user-images.githubusercontent.com/846565/228320351-dba87b96-8bf1-44db-8d56-3934bedecc68.png">

---

Depending on your tab procedure, and what is selected, it's sometimes possible to end up with a strange duplicate toolbar. Best described with a video:


https://user-images.githubusercontent.com/846565/228321914-935d22a6-f2bb-40f9-ae44-c2f78f934f73.mp4



---

I only see the new behavior in the post editor. I suspect we'll tackle the site editor seprately?

jasmussen commented:
Thanks so much for working on this, this is going to be hot, and a preferred default for many I'm sure! Functionally it's definitely getting there.

![status](https://user-images.githubusercontent.com/1204802/228445654-b4fcfd0f-3f64-44a8-8d7e-6a8322f2ea2c.gif)

Outside of the details Jay mention, I'll add a few:

The size of the plus button, and the spacing between the plus button and the tool selector regressed a bit compared to trunk:

<img width="335" alt="Screenshot 2023-03-29 at 08 28 27" src="https://user-images.githubusercontent.com/1204802/228445732-fa85c532-59ad-4952-bcc2-e4ca1d5722ba.png">
<img width="304" alt="Screenshot 2023-03-29 at 08 29 30" src="https://user-images.githubusercontent.com/1204802/228445738-805e7bd7-9ed0-428b-92a1-9c44e671960e.png">

It's bigger and there's no space. 

For awareness, there's https://github.com/WordPress/gutenberg/issues/46734 which details a "compact" button size of 32px (so inserter, toggled states, and publish buttons would all be 32px height) a la this:

<img width="1038" alt="Screenshot 2023-03-29 at 08 31 57" src="https://user-images.githubusercontent.com/1204802/228445975-daf5c05b-f44c-410f-83f5-0149cbd270d6.png">

I don't expect you to implement that in this branch, but let's be careful to at least not change dimensions from trunk in the mean time. 

Also, this dot should be the `·` mid-dot:

<img width="93" alt="Screenshot 2023-03-29 at 08 28 33" src="https://user-images.githubusercontent.com/1204802/228446129-d79d1871-1bb8-401c-b173-a5282492ed26.png">

Smaller, in any case.

draganescu committed:
- fixes regression on inserter size and margin
- restores fixed toolbar on mobile viewports
- adds hover states to toolbar toggles
- adds the correct level up icon
- moves CSS to the right files

Co-authored-by: Joen A. <1204802+jasmussen@users.noreply.github.com>
Co-authored-by: James Koster <846565+jameskoster@users.noreply.github.com>

draganescu committed:
fixes bogus toolbar on tabbing

draganescu committed:
tweak document toolbar margins

draganescu committed:
corrects separators

jasmussen commented on packages/block-editor/src/components/block-toolbar/style.scss:
```diff
@@ -20,6 +20,32 @@
+			font-size: 36px;
+			position: absolute;
+			left: $grid-unit-40 + $grid-unit-15;
+			bottom: math.div($grid-unit, 2);
```
You could probably SCSS this one, but we also have a handy $grid-unit-05.

jasmussen commented on packages/block-editor/src/components/block-toolbar/style.scss:
```diff
@@ -20,6 +20,32 @@
+		border: 0;
+		&::after {
+			content: "\00B7";
+			font-size: 36px;
```
I think the mid dot needs to be way smaller. In this design it's just the default 13px system font mid dot:

<img width="627" alt="Screenshot 2023-03-29 at 15 38 58" src="https://user-images.githubusercontent.com/1204802/228556351-053dd530-6d40-479c-a6cd-21859278e4eb.png">

jasmussen commented on packages/edit-post/src/components/header/header-toolbar/style.scss:
```diff
@@ -39,11 +39,91 @@
-	.edit-post-header-toolbar__left > .components-button.has-icon,
-	.edit-post-header-toolbar__left > .components-dropdown > .components-button.has-icon {
+	.edit-post-header-document-toolbar__left > .components-button.has-icon:first-child {
+		margin-right: 10px;
```
Can be $grid-unit-10 probably.

jasmussen replied:
Spacing wise, generally 16px ($grid-unit-20) and 8px ($grid-unit-10) are used. In this screenshot:

<img width="1079" alt="Screenshot 2023-03-29 at 15 43 18" src="https://user-images.githubusercontent.com/1204802/228557874-ecde46c1-270c-4959-88f3-5479aa59f4cf.png">

16px between blue plus button and row of controls, and 8px between every icon and separator.

jasmussen commented on packages/edit-post/src/components/header/header-toolbar/style.scss:
```diff
@@ -39,11 +39,91 @@
+
+	.edit-post-header-document-toolbar__left > .components-button.has-icon,
+	.edit-post-header-document-toolbar__left > .components-dropdown > .components-button.has-icon {
+		height: 32px;
```
32px can be $grid-unit-40, 4px can be $grid-unit-05.

draganescu replied:
Yes, for this one I specifically tweaked the px values to match perfectly what's in trunk. Using the grid unit values is 1px off when comparing. Will update using the grid units and check that out.

draganescu committed:
css tweaks to grid and also hiding toggles on tabbiing through

draganescu committed:
tweak to stop accidental tabbing into block tools

draganescu commented:
I've resolved the UI issues found by @jameskoster and @jasmussen . The remaining issue which I can't say how to solve so far is this:

- the block toolbar placed in the header _after_ the document toolbar, so that the document toolbar is still tabbable as per @alexstine 's [suggestion here](https://github.com/WordPress/gutenberg/pull/47277#issuecomment-1415926999) makes the block toolbar **not** the 1st tabbable element anymore, so Alt+F10 will move focus to the document toolbar. This is not what our tests expect.

This is a problem because when the toolbar is fixed or not the access to it via keyboard should be the same. This would work if we give up on having the document toolbar tabable into visibility, as the by removing it from the dom the block tools would be the 1st tabbable (esentially remove the `MaybeHide` component in this PR). But even then when the block toolbar is fixed the 1st item in it is the inserter not the block icon component like when the block toolbar is floating.

Another problem of similar nature is that when navigating with the keyboard we need to manage the focus when toggling between the two toolbars, @aaronrobertshaw or @tellthemachines  I'd love some help with how to handle these two problems as best as possible within the constraints. 🙏🏻 🙇🏻

scruffian commented:
For a tablet size, when I have a paragraph inside a group, the toolbar goes onto two lines:

<img width="788" alt="Screenshot 2023-03-30 at 13 52 14" src="https://user-images.githubusercontent.com/275961/228842232-e1d5be11-a401-48e5-bf22-bdc804d31558.png">

draganescu commented:
> The space between the 'up' arrow button and the separator is inconsistent with the other buttons. Looks like it because it's not inside a components-toolbar-group.

@jameskoster

alexstine commented:
@draganescu Just tested this again and you are right, this is a bit of a mess.
1. Can you explain to me visually what this top toolbar setting is doing? Is the goal to place the block toolbar where the document toolbar is to save space?
2. How do sighted users reveal the document toolbar when they wish to see it?
3. If you edit a block, shift+tab does not focus the block toolbar. You must pass through the buttons in previous tab stops first. Examples include options, settings, update/publish, preview, switch to draft, and finally you reach the toolbar. This is horrible UX for keyboard users but this is a rather obvious callout.
4. Unlabeled button: `<button type="button" id="id-c0l9if-2" tabindex="0" data-toolbar-item="true" class="components-button edit-post-header-toolbar__document-tools-toggle is-primary has-icon"></button>`
5. As you mentioned, alt+f10 no longer works to jump to the block toolbar.

We'll get there, thanks for the work on this.

tellthemachines commented:
> This is a problem because when the toolbar is fixed or not the access to it via keyboard should be the same.

Is this an actual accessibility issue, or is it just a problem because of how our tests are written? If the latter, it would be easier to change the tests 😅

alexstine commented:
@tellthemachines Accessibility issue for the reasons mentioned above. Shift+tab no longer places focus on the block toolbar either. If this is a visual change, it should not need to mess with focus/keyboard accessibility.

aaronrobertshaw commented:
>  @aaronrobertshaw I'd love some help with how to handle these two problems as best as possible within the constraints.

Thanks for the ping 👍 

After several weeks focused elsewhere and coming back to this PR having forgotten most context, I found the keyboard navigation really confusing. Due in large part to most of the issues @alexstine mentioned.

I'm struggling to see how we achieve the design which relocates several items, while keeping the same keyboard navigation regardless of whether or not the toolbar is fixed. I agree though, that it would be best to maintain the existing keyboard navigation flow for UI that already exists. 

> Currently users expect to be able to shift tab from the block to the block toolbar directly

This is an important point. I didn't realise how much I used this feature until it was no longer available.

Even if we accepted that because the fixed top toolbar isn't the default experience and needs never be enabled. It is hard to see it being a great experience without being able to quickly switch from the block back to its toolbar. Shift tab is the obvious choice for that, but could a dedicated keyboard shortcut be a compromise?

All of the above is a long way of saying I'm really short on ideas to move this forward, sorry.

@talldan has more experience in this area, it might be good if he can weigh in here.

afercia commented:
@draganescu thanks for the ping.

It's important to note that this PR breaks the expected keyboard interaction within the two toolbars. These are based on the ARIA toolbar design pattern. Behind the scenes, there are two `NavigableToolbar` components that provide the expected semantics and interaction. For more details on the breakage reasons, see a very similar case here: https://github.com/WordPress/gutenberg/issues/49479. In short: the toolbar buttons can't be hidden with CSS. Also rendering some of the buttons dynamically may break keyboard interaction. When the hidden/removed button happens to be the only focusable button expected within the toolbar, keyboard interaction is broken.

I also share all the accessibility concerns already reported by @alexstine and @joedolson so I'll avoid to repeat the same concerns.

## More general considerations
I went through all the feedback and comments in this PR and in the original issue https://github.com/WordPress/gutenberg/issues/40450. Lots of interesting explorations and thoughts there ❤️ 

However, seems to me this PR is going a bit too far. The original issues to address were:

1. The lack of parent selector for nested blocks
2. The overall increase in editor UI footprint

The initial suggested changes were:

1. **Add direct parent selector on the left.** Makes totally sense.
2. **Try center layout and/or floating docked toolbar.** OK, some better alignment can be surely tried.

Then, starting from https://github.com/WordPress/gutenberg/issues/40450#issuecomment-1285569461 the conversation evolved into something entirely different: morphing the document tools into block tools on selection. Which basically means a unified toolbar.

I'd like to remind that before WordPress 5.6 the top bar used to be a 'unified' toolbar already. Screenshot from WordPress 5.5:

<img width="1107" alt="wp55" src="https://user-images.githubusercontent.com/1682452/229149329-038e8f1f-b117-400f-ae48-f4c7a59badc3.png">

It was than changes to two separate toolbars, and for good reasons. The split in two was discussed by the design team, see [the original G2 issue](https://github.com/WordPress/gutenberg/issues/20592#issuecomment-654983754), and having the two toolbars in two separate rows was the preferred option. See also some a11y considerations here: https://github.com/WordPress/gutenberg/issues/20592#issuecomment-700700478. Other important considerations were pointed out in the related pull requests, see https://github.com/WordPress/gutenberg/pull/31134 and https://github.com/WordPress/gutenberg/pull/30234.

This PR would bring us back to a unified toolbar, re-introducing most of the problems addressed two years ago.

On top of that, this PR aims to automatically toggle the toolbars visibility depending on the selection state and keyboard interaction. As already pointed out in this PR and in several issues and PRs across the history of this project, 'disappearing and appearing elements' are always a barrier to non-sighted users. I'd add they're always confusing also for sighted users, as they add cognitive load, trigger loss of visual context, and potentially hide important tools thus breaking the user's flow.

To me, that's already a good reason to rethink the impact of this PR and limit the changes to the ones that were originally proposed:
- add the parent selector
- try a better alignment

There's a lot of additional usability / accessibility considerations that make me think the proposed unified toolbar is less than ideal. Trying to avoid a [TL;DR](https://www.urbandictionary.com/define.php?term=tl%3Bdr), here's some of the main concerns:

- As said previously, the expected keyboard interaction of the ARIA toolbar pattern is now broken.
- Re: this comment: [I'm hesitant to bury the List view control](https://github.com/WordPress/gutenberg/issues/40450#issuecomment-1355019768). That's a very good point. I'd say burying the Undo and Redo buttons is even more impactful though. As pointed out, the List view is typically ued only a few times in a typical user flow. However, Undo and Redo are likely used way more often. I'm not sure forcing users to perform and extra click each time they need to Undo or Redo is an improvement. Yes, there are keyboard shortcuts but also yes: most users will look for the Undo and Redo buttons.
- The 'Up' and 'selected block' buttons to switch toolbars are very unclear to me. Not intuitive at all. Aside: right now these two buttons are unlabelled and miss a tooltip.
- The 'selected block' button performs always the same action but its icon changes depending on the selected block. By its ever changing look, I'm not sure I would be able to understand what this button does. It always confuses me.
- I'd argue to whole concept of 'levels', where the Document tools is the 'top level' toolbar and the Block tools is the 'second level' toolbar is too much to be easily understandable for users. As a user, I'd just want the toolbar I need back into view. I wouldn't want to be forced to think each time at where I am now and whether I need to go 'up' or 'down'.
- Pressing Enter on the 'Up' and 'selected block' buttons triggers a focus loss, as they get removed from the DOM while they are focused.
- As pointed out by @alexstine, For screen reader users who are navigating in virtual mode, there is now no longer a way to access the document tools toolbar since it will be hidden in the DOM/conditionally removed. This is a serious barrier.
- Minor but still a problem: when the Block tools toolbar is shown, the Top bar will show two 'vertical ellipsis' buttons that are visually placed in the same row, both labelled 'Options'. Arguably a good UI. Screenshot:

<img width="1283" alt="two options same row" src="https://user-images.githubusercontent.com/1682452/229159742-f2c72a1f-6a51-4f56-b3b5-81fa7c5fb0a8.png">

With the current split toolbars, at least it's more clear they belong to two different contexts:

<img width="1278" alt="Screenshot 2023-03-31 at 17 14 19" src="https://user-images.githubusercontent.com/1682452/229160745-fa9bec19-b46f-44b8-9441-4b1324e81974.png">

- Last but not least: in the Site Editor there's way less space for a unified toolbar because of the `DocumentActions` component displayed at the center of the top bar,

## Conclusion
I'm not sure this PR would be an actual improvement, as is. To me, the disappearing / appearing elements and hiding important tools are a no starter IMHO.

However, I wouldn't be opposed to try the unified toolbar as long as the Document tools stay always visible. No disappearing / appearing. No hidden tools. From there, I'm pretty sure we could refine and address all the remaining concerns.

I do realize that would lead us back to the lack of space issue though. As I see it, that's more a design issue. The fixed height ot the top bae and its absolute positioning are part of the problem. The very nature of content on the web is that it should dreflow as necessary depending on the viewport and the device in use. Trying to 'squeeze' or hide the content just to make it fit into an UI that doesn't provide enough space just proves the UI isn't design around the content.

alexstine commented:
Yeah, I am not a fan of the appearing/disappearing DOM elements either but it seems to be the way of the future with React/Vue/other frameworks. Not the largest issue in the world but things would be less complicated if things did not disappear from the DOM. Something a lot of times sighted people and designers take for granted how hard these interactions can be for people without visual indicators. I totally support conditional rendering for dialogs but it is way over used in Gutenberg even outside of this PR.

draganescu commented:
I have tried but not succeeded in a different approach, yet I still tink it's possible:

- make the block toolbar in fixed mode render in a place where:
  - it is possible that we position it absolutely on top of the editor header, including z-index
  - it still is the nearest toolbar (Alt+f10 and shift+tab work)
- use an observer to hide it if the focus is not in canvas

Everything else stays the same for mouse users with buttons that can show and hide document and block tools.

My challenge was where to place the block toolbar in fixed mode to respect the bullet requirements above.

draganescu committed:
attempt to solve both ARIA toolbar access and tab stop to document tools

draganescu commented:
Ok, here is the current state: playing around with the DOM position and some heavy focus management, I can achieve the state in the video. 


https://user-images.githubusercontent.com/107534/229874234-c41560bc-2722-4b06-b328-59e0aa3e4d65.mp4

There still is work to do:

- I have an effect that is too eager and moves the focus to the toolbar when selection changes between document and block
- For some reason Alt+F10 focuses document tools which hides the block toolbar. But if I don't listen to the focus event Alt+F10 focuses block tools, I swear it's playing quantum jokes on me.

alexstine commented:
@draganescu It is working much better. Keep going, you are close.

Thanks.

draganescu commented:
TL;DR; there appears to be a way to not break the keyboard interaction at all but it has sent me basically to square one.

With the help of @MaggieCabrera and @jeryj  I have found a way to reuse the fixed block toolbar and not have to change where it is rendered. Instead, I can simply reposition it on top of the document tools and style it to fit visually there. This has the great advantage that all the keyboard interaction remains the same! No more tricks.

However it has some challenges which I am looking into:

- the buttons that toggle the document tools and the block tools can no longer use a simple state to toggle between toolbars because one toolbar is in header and the other one is in a popover somewhere in content. It is dubious both to use a store and to drill down props (because the button that toggles the document tools is in the block toolbar component which is way too many layers to drill down props and it's also in another package).
  - to this a solution may be to put both the toggling buttons in the header and use CSS to make a nice puzzle, and use the edit post store to store toolbar visibility, but it's still weird.
- there'll be a lot of CSS to guard small viewports from the desktop changes
- I still have to use the `:focus-within` trick on document tools to hide block tools, but, not sure I can via CSS so it'll probably be a focus listener.

---

As far as the approach in this PR, Alt+F10 is handled by a hook that manages focus when this keyboard shortcut is used. This hook is not specific to block toolbar but to all navigable toolbars. Because this PR puts two navigable toolbars in the same container Alt+F10 always selects the 1st which is document tools. The other thing I still have to see if I can change this in a reasonable way (unlikely).

draganescu commented:
Closing in favor of #49634
